readme.txt
======================
OSCAR McMaster Project
======================

------------
Requirements  
------------
These are not necessarily requirements but the version of software used by the author of this read me at the time of this writing. Generally speaking any newer version should work.
- mysql 5.1.52
- jdk1.6.0_23
- tomcat-6.0.29
- maven 2.2.1
- ant 1.8.1

-----------
Directories
-----------
catalina_base : is a catalina_base that's provided, it can be useful as a starting point for deployments or development.
database : contains sql scripts to initialise the database schema.
docs : developer style documentation
local_repo : a maven repository for libraries not found at the publicly accessible maven repositories.
src : the source code for this project, structured in a standard maven structure.
utils : some random utilities and files we don't have anywhere else to put.

--------
Building 
--------
This is a standard maven project. "mvn package" should create a target directory and there should be a war file in there.

To test jsp compilations "ant -f jspc.xml" can be run, it presumes you have just run "mvn package" and that the target directory and corresponding war file is present.

-------------
NetBeans Note
-------------
Add this to a netbeans build.xml file for it to build and run.

<target name="-post-compile">
        <echo message="deleting hbm.xml files from src directory." />
            <delete>
                <fileset dir="${build.classes.dir}/src/" includes="**/*.hbm.xml"/>
            </delete>
</target>
    


changelog.txt
2009-08-25 : Generic Intakes: added support for internal labels, and the loading of form specific javascript
2009-08-25 : Support for disease registry associations, automatch, and auto-insert for certain issues.
2009-05-13 : Support for digital signature pads. (Starting with caisi integrator consent) (Ted)
2009-04-28 : Expand form edited date to include time. (Dan Brow)
2009-04-28 : Fixed "Abnormal" reporting for GDML labs (DavidD)
2009-04-28 : Added support for HL7 labs from ICL (Intercounty Medical Labs) (DavidD)
2009-04-28 : Added searchable "Remit To" in 3rd Party ON Billing (DavidD)
2009-02-28 : Added ability to link a Document or Lab with a Tickler
2009-01-28 : Changed how to edit existing notes in new encounter.  Click on edit link instead of note (Randy)
             Pass demographic numbers in post of form rather than in url for prevention report(Randy)
2008-12-31 : client image displays on client summary page (Ted)
2008-12-07 : First commit for quatro merge
	     Includes most of the security framework, homepage, ORGtree viewer
	     Relevant db changes (not yet incorporated into main files until we do some testing)
2008-11-18 : Caisi Integrator also provides photos when client is copied from remote server.
2008-11-17 : removed the old RFQ required intake fields editor.

-----------------
Branch 2008-11-03
-----------------

2008-10-25:  Added Limited Use Code Integration to Rx (Jay)
2008-10-20 : Added ability to delete/undelete eDocs and filter display of eDocs (Randy,2032846)
2008-10-09 : Added Photo as an option in DMS, don't use update-2008-10-09.sql if you already have this option in your drop downi list. (Dan Brow)
2008-10-07 : Increased billing unit field size to 6 characters to accommadate percentage billing. (Dan Brow)
2008-09-26 : fix link to encounter in myOSCAR and added code to insert myOSCAR message into encounter note (Randy)
2008-09-25 : new provider service reports export (Ted)
2008-09-23 : ability to bulk export activity reports to csv (Ted)
2008-09-22 : When editing appointment only grab status from db if page is loading for first time else grab it from request (Randy)
2008-09-16 : add/edit appointment page, change input field for status to dropdown list box (Toby)
2008-09-12 : Caisi Integrator has the ability to manage which demographics are linked together. (Ted)
2008-09-11 : Fixed error when correcting invoices duplicate codes can be created for ON billing (Randy,2105933)
2008-09-09 : Fixed bug with Report By Template with dates in Firefox 3 and Safari which only the bottom date param would work (Jay,2102612)
             Fixed bug with Clinic Admin page that would log the user out if they clicked on the page (Jay,2102572)
2008-09-07 : Added practitioner number field in provider settings. Added Practitioner number to signature of Rx pad. (Dan)
2008-09-06 : Changed current prescription description in consultation from the short version to the long version. (Jay,2097084)
2008-09-05 : Caisi Integrator now shares prevention data across facilities (Ted)
2008-09-03 : Add copy favorite prescriptions function (Toby)
2008-09-02 : Changed icd10 table to be supported by desease registry
2008-09-02 : Added support for adding caisi issues set as 'major' to desease registry
2008-08-27 : Added newsletter option to demographic (Randy)
2008-08-26 : Tagged repositiory with  B4_MULTIPLE-ECHARTS (Randy)
             Changed cme/encounter to work off of request scope forms in
             both CAISI and OSCAR  
2008-08-26 : Paste oscar messages into new encounter (Randy)             
2008-08-22 : Old CME has option to sort asc and desc now. (Ted)
2008-08-01 : detailed history of caisi integrator consent is viewable. (Ted)
2008-07-28 : caisi integrator uses ssl. (Ted)
2008-07-22 : caisi integrator supports referrals between facilities. (Ted)

---------------------------
Changes prior to 2008-07-17
---------------------------
PMM:

-Home and Logout buttons are now in the top right hand corner

-“Oscar Medical” the link used to bring you back to case management screen has been renamed, the link on the navigator bar is now called “Case Management”

-New Report link: User Created Form Report: This will create a report based on a specific user created form

-Dependencies (spouse, partner, child, sibling, etc.): With the upgrade comes the new feature that allows a provider to link more then one file together. For example, a woman and her 2 children are staying in a shelter; the provider can create separate files for all 3 individuals, but can link them based on their relationship. Follow these steps:
-Program Management Module
-Summary Page for a client
-Second Tab down says “family”
-Click on update and enter in the last name to search for family members/dependents
-Add member to client file

-Health and Safety Message on client PMM Summary: On the summary tab for a client, in the Program Management Module, users can add in a health and safety message simply by clicking on “New Health and Safety”, entering message and pressing save

-Integrator Consent: This consent is in place to indicate the level of information sharing a client would like (ranging from no information shared to all information shared). This will not be in use until the integrator is functional and if Seaton House agrees to participate in integration. 


-Addition of the concept of facilities: A facility is a larger entity that holds together smaller programs. For example, the City of Toronto would be a Facility and the agencies would be the City run shelters. For the purposes of Seaton House, Seaton House will be the Facility.  


-Registration Editor: This would allow for the customization of questions that must be mandatory in the intake. This will not need to be used by Seaton House Administration.
-Intake Editor: This is the ability to customize an intake (adding questions, removing questions, etc). This will not need to be used by Seaton House Administration


web/js/IntakeB.js


function redirectToClientSearch(demographicNo,startRowCount,path)
{
	document.intakeBForm.action= path + "?searchForClient=N";
	document.intakeBForm.submit();
}


    function onPrint() {
//        document.forms[0].submit.value="print";
//        var ret = checkAllDates();
//        if(ret==true)
//        {
//            ret = confirm("Do you wish to save this form and view the print preview?");
//        }
//        return ret;
        window.print();
    }
    function onSave() 
	{
        document.forms[0].submit.value="save";
        var ret = checkAllDates();
		
//alert("onSave: checkAllDates() = " + ret);		
		if(ret)
		{
			ret = checkAllIntegers();
		}
        if(ret)
        {
            ret = confirm("Are you sure you want to save this form?");
        }
        return ret;
		
    }
    
    function onUpdate(path) 
	{
alert("onUpdate: just entered");		
	
        document.forms[0].submitForm.value="update";
        var ret = checkAllDates();
        
alert("onUpdate: checkAllDates() = " + ret);		
        
		if(ret)
		{
			ret = checkAllIntegers();
		}
		
alert("onUpdate: checkAllIntegers() = " + ret);		
		
        if(ret)
        {
            ret = confirm("Are you sure you want to update this form?");
            if(ret)
            {
            	document.forms[0].action = path+"?actionType=update";
            	
            	document.forms[0].submit();
            	
            }
        }
    }
    
    function onExit() {
        if(confirm("Are you sure you wish to exit without saving your changes?")==true)
        {
            window.close();
        }
        return(false);
    }
    function onSaveExit() {
        document.forms[0].submit.value="exit";
        var ret = checkAllDates();
		if(ret)
		{
			ret = checkAllIntegers();
		}	
        if(ret)
        {
            ret = confirm("Are you sure you wish to save and close this window?");
        }
		return ret;
    }

    function popupPage(varpage) { //open a new popup window
      var page = "" + varpage;
      windowprops = "height=600,width=700,location=no,scrollbars=yes,menubars=no,toolbars=no,resizable=yes,screenX=50,screenY=50,top=20,left=20";
      var popup=window.open(page, "printlocation", windowprops);
      if (popup != null) {
        if (popup.opener == null) {
          popup.opener = self;
        }
      }
    }

/**
 * DHTML date validation script. Courtesy of SmartWebby.com (http://www.smartwebby.com/dhtml/)
 */
// Declaring valid date character, minimum year and maximum year
var dtCh= "/";
var minYear=1900;
var maxYear=3100;

    function isInteger(s)
	{
        var i;
//alert("isInteger: s =" + s);			
		
        for (i = 0; i < s.length; i++)
		{
            // Check that current character is number.
            var c = s.charAt(i);
            if (((c < "0") || (c > "9")))
			{
				return false;
			}
        }
        // All characters are numbers.
        return true;
    }

    function isCurrency(s)
	{
        var i;
//alert("isInteger: s =" + s);			
		
        for (i = 0; i < s.length; i++)
		{
            // Check that current character is number.
            var c = s.charAt(i);
//alert("isInteger: c[" + i + "] =" + c);			
            if (  ((c < "0") || (c > "9"))  &&  c != "."  )
			{
				return false;
			}
        }
        // All characters are numbers.
        return true;
    }

    function stripCharsInBag(s, bag){
        var i;
        var returnString = "";
        // Search through string's characters one by one.
        // If character is not in bag, append to returnString.
        for (i = 0; i < s.length; i++){
            var c = s.charAt(i);
            if (bag.indexOf(c) == -1) returnString += c;
        }
        return returnString;
    }

    function daysInFebruary (year){
        // February has 29 days in any year evenly divisible by four,
        // EXCEPT for centurial years which are not also divisible by 400.
        return (((year % 4 == 0) && ( (!(year % 100 == 0)) || (year % 400 == 0))) ? 29 : 28 );
    }
    function DaysArray(n) {
        for (var i = 1; i <= n; i++) {
            this[i] = 31
            if (i==4 || i==6 || i==9 || i==11) {this[i] = 30}
            if (i==2) {this[i] = 29}
       }
       return this
    }

    function isDate(dtStr){
        var daysInMonth = DaysArray(12)
        var pos1=dtStr.indexOf(dtCh)
        var pos2=dtStr.indexOf(dtCh,pos1+1)
        var strMonth=dtStr.substring(0,pos1)
        var strDay=dtStr.substring(pos1+1,pos2)
        var strYear=dtStr.substring(pos2+1)
        strYr=strYear
        if (strDay.charAt(0)=="0" && strDay.length>1) strDay=strDay.substring(1)
        if (strMonth.charAt(0)=="0" && strMonth.length>1) strMonth=strMonth.substring(1)
        for (var i = 1; i <= 3; i++) {
            if (strYr.charAt(0)=="0" && strYr.length>1) strYr=strYr.substring(1)
        }
        month=parseInt(strMonth)
        day=parseInt(strDay)
        year=parseInt(strYr)
        if (pos1==-1 || pos2==-1){
            return "format"
        }
        if (month<1 || month>12){
            return "month"
        }
        if (day<1 || day>31 || (month==2 && day>daysInFebruary(year)) || day > daysInMonth[month]){
            return "day"
        }
        if (strYear.length != 4 || year==0 || year<minYear || year>maxYear){
            return "year"
        }
        if (dtStr.indexOf(dtCh,pos2+1)!=-1 || isInteger(stripCharsInBag(dtStr, dtCh))==false){
            return "date"
        }
    return true
    }


    function valDate(dateBox)
    {
        try
        {
            var dateString = dateBox.value;
            if(dateString == "")
            {
    //            alert('dateString'+dateString);
                return true;
            }
            var dt = dateString.split('/');
            var y = dt[0];
            var m = dt[1];
            var d = dt[2];
            var orderString = m + '/' + d + '/' + y;
            var pass = isDate(orderString);

            if(pass!=true)
            {
                alert('Invalid '+pass+' in field ' + dateBox.name);
                dateBox.focus();
                return false;
            }
        }
        catch (ex)
        {
            alert('Catch Invalid Date in field ' + dateBox.name);
            dateBox.focus();
            return false;
        }
        return true;
    }



	function checkAllIntegers()
	{
        var ret = true;
/*		
        if(!isInteger(document.forms[0].year.value))
		{
	        alert ("You must type in a number in the field.");
		
			document.forms[0].year.value = "";
			document.forms[0].year.focus();
            ret = false;
        }
		else 
*/		
		if(!isInteger(document.forms[0].drinksPerDay.value) )
		{
	        alert ("You must type in a number in the field.");
		
			document.forms[0].drinksPerDay.value = "";
			document.forms[0].drinksPerDay.focus();
		
            ret = false;
        }
		else if(!isInteger(document.forms[0].drinksPerWeek.value))
		{
	        alert ("You must type in a number in the field.");
		
			document.forms[0].drinksPerWeek.value = "";
			document.forms[0].drinksPerWeek.focus();
		
            ret = false;
        }
		else if(!isInteger(document.forms[0].drinksPerMonth.value))
		{
	        alert ("You must type in a number in the field.");
		
			document.forms[0].drinksPerMonth.value = "";
			document.forms[0].drinksPerMonth.focus();
		
            ret = false;
        }
		else if(!isInteger(document.forms[0].yearsOfEducation.value))
		{
	        alert ("You must type in a number in the field.");
		
			document.forms[0].yearsOfEducation.value = "";
			document.forms[0].yearsOfEducation.focus();
		
            ret = false;
        }
		else if(!isInteger(document.forms[0].howLongEmployed.value))
		{
	        alert ("You must type in a number in the field.");
		
			document.forms[0].howLongEmployed.value = "";
			document.forms[0].howLongEmployed.focus();
		
            ret = false;
        }
		else if(!isInteger(document.forms[0].howLongUnemployed.value))
		{
	        alert ("You must type in a number in the field.");
		
			document.forms[0].howLongUnemployed.value = "";
			document.forms[0].howLongUnemployed.focus();
		
            ret = false;
        }
		else if(!isCurrency(document.forms[0].amtOwing.value))
		{
	        alert ("You must type in a currency amount in the field.");
		
			document.forms[0].amtOwing.value = "";
			document.forms[0].amtOwing.focus();
		
            ret = false;
        }
		else if(!isCurrency(document.forms[0].howMuchYouReceive.value))
		{
	        alert ("You must type in a currency amount in the field.");
		
			document.forms[0].howMuchYouReceive.value = "";
			document.forms[0].howMuchYouReceive.focus();
		
            ret = false;
        }
		
        return ret;
	}

    function checkAllDates()
    {
        var b = true;
        if(valDate(document.forms[0].dateAssessment)==false)
		{
            b = false;
        }
		else if(valDate(document.forms[0].dateEnteredSeaton)==false)
		{
            b = false;
        }
		else if(valDate(document.forms[0].dateExitedSeaton)==false)
		{
            b = false;
        }
		else if(valDate(document.forms[0].dateLastDoctor1Contact)==false)
		{
            b = false;
        }
		else if(valDate(document.forms[0].dateLastDoctor2Contact)==false)
		{
            b = false;
        }
		else if(valDate(document.forms[0].dateLivedThere)==false)
		{
            b = false;
        }
		else if(valDate(document.forms[0].dateLastContact1)==false)
		{
            b = false;
        }
		else if(valDate(document.forms[0].dateLastContact2)==false)
		{
            b = false;
        }
		else if(valDate(document.forms[0].dateLastContact3)==false)
		{
            b = false;
        }
		else if(valDate(document.forms[0].dateLastContact4)==false)
		{
            b = false;
        }
		else if(valDate(document.forms[0].whenMadeAppforOtherIncome)==false)
		{
            b = false;
        }

        return b;
    }
	
	function clickCheckBox(fieldObj)
	{
		if(fieldObj.checked)
		{
			fieldObj.checked = false;
		}
		else if(!fieldObj.checked)
		{
			fieldObj.checked = true;
		}
	}	
	function clickRadio(fieldObj)
	{
		if(fieldObj.checked)
		{
			fieldObj.checked = false;
		}
		else if(!fieldObj.checked)
		{
			fieldObj.checked = true;
		}
	}	
	
